sdk/iot/core/inc/az_iot_hub_client.h

```diff
@@ -0,0 +1,505 @@
+ *                          filter.
+ * @param out_mqtt_topic_filter The output #az_span containing the MQTT topic filter.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_subscribe_topic_filter_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_topic_filter,
+    az_span* out_mqtt_topic_filter);
+
+/**
+ * @brief A method request received from IoT Hub.
+ *
+ */
+typedef struct az_iot_hub_client_method_request
+{
+  az_span request_id; /** @brief The request id.
+                       * @note The application must match the method request and method response. */
+  az_span name; /** @brief The method name. */
+} az_iot_hub_client_method_request;
+
+/**
+ * @brief Attempts to parse a received message's topic.
+ *
+ * @param client
+ * @param received_topic An #az_span containing the received topic.
+ * @param out_request If the message is a method request, this will contain the
+ *                    #az_iot_hub_client_method_request.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_received_topic_parse(
+    az_iot_hub_client const* client,
+    az_span received_topic,
+    az_iot_hub_client_method_request* out_request);
+
+/**
+ * @brief Gets the MQTT topic that must be used to respond to method requests.
+ *
+ * @param client
+ * @param request_id The request id. Must match a received #az_iot_hub_client_method_request
+ *                   request_id.
+ * @param status The status. (E.g. 200 for success.)
+ * @param mqtt_topic An empty #az_span with sufficient capacity to hold the MQTT topic.
+ * @param out_mqtt_topic The output #az_span containing the MQTT topic.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_hub_client_methods_response_publish_topic_get(
+    az_iot_hub_client const* client,
+    az_span request_id,
+    uint16_t status,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic);
+
+/**
+ *
+ * Twin APIs
+ *
+ */
+
+/**
+ * @brief Azure IoT Hub status codes.
+ *
+ */
+typedef enum
+{
+  // Service success codes
+  AZ_IOT_HUB_CLIENT_STATUS_OK = 200,
+  AZ_IOT_HUB_CLIENT_STATUS_ACCEPTED = 202,
+  AZ_IOT_HUB_CLIENT_STATUS_NO_CONTENT = 204,
+
+  // Service error codes
+  AZ_IOT_HUB_CLIENT_STATUS_BAD_REQUEST = 400,
+  AZ_IOT_HUB_CLIENT_STATUS_UNAUTHORIZED = 401,
+  AZ_IOT_HUB_CLIENT_STATUS_FORBIDDEN = 403,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_FOUND = 404,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_ALLOWED = 405,
+  AZ_IOT_HUB_CLIENT_STATUS_NOT_CONFLICT = 409,
+  AZ_IOT_HUB_CLIENT_STATUS_PRECONDITION_FAILED = 412,
+  AZ_IOT_HUB_CLIENT_STATUS_REQUEST_TOO_LARGE = 413,
+  AZ_IOT_HUB_CLIENT_STATUS_UNSUPPORTED_TYPE = 415,
+  AZ_IOT_HUB_CLIENT_STATUS_THROTTLED = 429,
+  AZ_IOT_HUB_CLIENT_STATUS_CLIENT_CLOSED = 499,
+  AZ_IOT_HUB_CLIENT_STATUS_SERVER_ERROR = 500,
+  AZ_IOT_HUB_CLIENT_STATUS_BAD_GATEWAY = 502,
+  AZ_IOT_HUB_CLIENT_STATUS_SERVICE_UNAVAILABLE = 503,
+  AZ_IOT_HUB_CLIENT_STATUS_TIMEOUT = 504,

```

Comment:
>AZ_IOT_HUB_CLIENT_STATUS_TIMEOUT [](start = 2, length = 32)

Bigger question I know, but what happens if we need to add a new status code here, in terms of versioning of enum?

Reply:
Why would that be a problem? It's additive and we don't support binary app-compat.
Existing apps will continue to compile, new apps can take a dependency.

Reply:
I see.  Only problem if an app does like a switch() without a default, which they shououldn't do anyway.